### PR TITLE
Fix: Crash when search encounters a 404 error 

### LINKED ISSUE #4411: Spotify singleton import does not continue after encountering a 404 error
I sometimes run into this error when I import tracks as singletons and a 404 is encountered. My expectation was that the import will continue with the next track, but that is not happening and the import process just exits. 
I just modified the API error to include more details and here is the trace:

```
/home/arsaboo/shared/music/Coke Studio @ MTV Season 2 (2012)/Coke Studio @ MTV Season 1 (2011) - Dheere Dheere.flac
Tagging track: Megha Sriram Dalton - Dheere Dheere
URL:
    https://musicbrainz.org/recording/efc2a2cd-9ed8-4436-8007-0ebe1a166873
(Similarity: 100.0%)
[A]pply, More candidates, Skip, Use as-is, Enter search, enter Id, aBort,
eDit, edit Candidates? a
Traceback (most recent call last):
  File "/home/arsaboo/.local/bin/beet", line 8, in <module>
    sys.exit(main())
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/ui/__init__.py", line 1304, in main
    _raw_main(args)
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/ui/__init__.py", line 1291, in _raw_main
    subcommand.func(lib, suboptions, subargs)
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/ui/commands.py", line 1034, in import_func
    import_files(lib, paths, query)
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/ui/commands.py", line 974, in import_files
    session.run()
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/importer.py", line 340, in run
    pl.run_parallel(QUEUE_SIZE)
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/util/pipeline.py", line 446, in run_parallel
    raise exc_info[1].with_traceback(exc_info[2])
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/util/pipeline.py", line 314, in run
    for msg in _allmsgs(out):
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/util/pipeline.py", line 467, in pull
    out = coro.send(msg)
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/util/pipeline.py", line 193, in coro
    func(*(args + (task,)))
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/importer.py", line 1376, in lookup_candidates
    task.lookup_candidates()
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/importer.py", line 913, in lookup_candidates
    prop = autotag.tag_item(self.item, search_ids=self.search_ids)
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/autotag/match.py", line 518, in tag_item
    for track_info in hooks.item_candidates(item, search_artist, search_title):
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/plugins.py", line 573, in decorated
    for v in generator(*args, **kwargs):
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/autotag/hooks.py", line 647, in item_candidates
    yield from plugins.item_candidates(item, artist, title)
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/plugins.py", line 392, in item_candidates
    yield from plugin.item_candidates(item, artist, title)
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/plugins.py", line 749, in item_candidates
    tracks = self._search_api(
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beetsplug/spotify.py", line 400, in _search_api
    self._handle_response(
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beetsplug/spotify.py", line 197, in _handle_response
    raise SpotifyAPIError("API Error: {}\nURL: {}\nparams: {}".
beetsplug.spotify.SpotifyAPIError: API Error: 404
URL: https://api.spotify.com/v1/search
params: {'q': 'Jiya Laage Na artist:Shankar Mahadevan, Khagan Gogoi, Megha Sriram Dalton, Akriti Kakar & Leslie Lewis', 'type': 'track'}
arsaboo@arsplex:~$

```
Any thoughts as to what may be going on and I can try to fix the issue?

### PR DESCRIPTION
**Pull request by arsaboo**
Fixes #4411

I had to add a few additional guard clauses to prevent errors. 

Tested this fix with the same track with both Spotify and MB match and it works fine.

### CODE REVIEW

**Comment by sampsyo** [comment]
Hi! I'm not entirely sure why the `plugins.py` changes are necessary here. It looks like the main change in the plugin is that `_search_api` can now return `None`, but the only place where that function is called is here:
https://github.com/beetbox/beets/blob/7209c5f922ad7bcc9ae88991e1c5cf5494a02577/beetsplug/spotify.py#L524-L527

…which already has a handler for the case where the return value is empty. Any chance you could explain a little more what necessitated the extra `is not None` checks?

**Comment by arsaboo** [comment]
Without those `None` checks in the plygins.py, we get `NoneType` errors. 
You can try running the code until this [85e58d4](https://github.com/beetbox/beets/pull/4413/commits/85e58d48a26463f0f370f5f051f65bb9451431af) commit and you will see those errors. Consider the following: https://github.com/beetbox/beets/blob/7209c5f922ad7bcc9ae88991e1c5cf5494a02577/beets/plugins.py#L752

My guess is that some of the other plugins (maybe MB) return `None`, causing the error. The same is the case in other places. 
In any case, I don't see a downside in adding those checks.

**Comment by sampsyo** [comment]
Is there any chance you could include a traceback? Even if the MusicBrainz backend sometimes returns None, I am not seeing how that would be affected by the current change… it would be great to understand what's going on.

**Comment by arsaboo** [comment]
Here's the trace with plugins.py reverted:
```
Traceback (most recent call last):
  File "/home/arsaboo/.local/bin/beet", line 8, in <module>
    sys.exit(main())
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/ui/__init__.py", line 1304, in main
    _raw_main(args)
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/ui/__init__.py", line 1291, in _raw_main
    subcommand.func(lib, suboptions, subargs)
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/ui/commands.py", line 1034, in import_func
    import_files(lib, paths, query)
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/ui/commands.py", line 974, in import_files
    session.run()
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/importer.py", line 340, in run
    pl.run_parallel(QUEUE_SIZE)
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/util/pipeline.py", line 446, in run_parallel
    raise exc_info[1].with_traceback(exc_info[2])
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/util/pipeline.py", line 314, in run
    for msg in _allmsgs(out):
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/util/pipeline.py", line 467, in pull
    out = coro.send(msg)
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/util/pipeline.py", line 193, in coro
    func(*(args + (task,)))
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/importer.py", line 1376, in lookup_candidates
    task.lookup_candidates()
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/importer.py", line 913, in lookup_candidates
    prop = autotag.tag_item(self.item, search_ids=self.search_ids)
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/autotag/match.py", line 518, in tag_item
    for track_info in hooks.item_candidates(item, search_artist, search_title):
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/plugins.py", line 573, in decorated
    for v in generator(*args, **kwargs):
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/autotag/hooks.py", line 647, in item_candidates
    yield from plugins.item_candidates(item, artist, title)
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/plugins.py", line 392, in item_candidates
    yield from plugin.item_candidates(item, artist, title)
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/plugins.py", line 752, in item_candidates
    return [self.track_for_id(track_data=track) for track in tracks]
TypeError: 'NoneType' object is not iterable
```
If I fix that we get error on another line.

File: beetsplug/spotify.py [hunk 1]
```diff
@@ -395,15 +396,17 @@ def _search_api(self, query_type, filters=None, keywords=''):
...
                 requests.get,
                 self.search_url,
                 params={'q': query, 'type': query_type},
             )
-            .get(query_type + 's', {})
-            .get('items', [])
-        )
+        except SpotifyAPIError as e:
+            self._log.debug('Spotify API error: {}', e)
+            return None
```
Comment (sampsyo) [inline beetsplug/spotify.py:407]: This should return an empty sequence, i.e., `[]`.

**Review by sampsyo — CHANGES_REQUESTED** [review]
Aha; sorry, I feel dumb! I didn't realize that `_search_api` was called within the `MetadataSourcePlugin` base class:
https://github.com/beetbox/beets/blob/7209c5f922ad7bcc9ae88991e1c5cf5494a02577/beets/plugins.py#L732

We will want this plugin to return an _empty list of results_ instead of `None`. That should avoid the `None` special-casing.

**Comment by arsaboo** [comment]
The empty sequence helps....it works now 👍

**Review by sampsyo — APPROVED** [review]
LGTM; thanks!

**Pull request merged**